pkg/secrethub/client.go

```diff
@@ -19,15 +20,29 @@ type Client interface {
 	Users() UserService
 }
 
+// Decryptor decrypts data, typically an account key.
+type Decryptor interface {
+	// Unwrap decrypts data, typically an account key.
+	Unwrap(ciphertext crypto.CiphertextRSAAES) ([]byte, error)
+}
+
+// Encryptor encrypts data, typically an account key.
+type Encryptor interface {
+	// Fingerprint returns an identifier by which the server can identify the credential, e.g. a username of a fingerprint.
+	Fingerprint() (string, error)
+	// Wrap encrypts data, typically an account key.
+	Wrap(plaintext []byte) (crypto.CiphertextRSAAES, error)
+}
+
 type clientAdapter struct {
 	client client
 }
 
 // NewClient creates a new SecretHub client.
 // It overrides the default configuration with the options when given.
-func NewClient(credential Credential, opts *ClientOptions) Client {
+func NewClient(decryptor Decryptor, encryptor Encryptor, signer auth.HTTPSigner, opts *ClientOptions) Client {
```

Comment:
How would this work if there is no `Encryptor` available in the client? For example, in the case of a service using KMS, the service does not necessarily have the permission to encrypt using KMS.

Do we provide an interface that returns an error or should we remove the `Encryptor` from this signature?

Reply:
I like the idea of providing an interface that returns an error. When implementing it like that, the client doesn't have to check for existence of an encryptor/decryptor at every usage. It even is unaware that encryptor/decryptors can be omitted at all. The message of the error is determined by the user of the client. This is where we know what message to return. We don't have to check for the type of the returned error by the client.

---

Alternatively we could do:
```
client := secrethub.NewClient(signer).Decryptor(decryptor).Encryptor(encryptor)
```

The `Decryptor` and `Encryptor` funcs set the corresponding fields and return a pointer to the client.
When using a func that requires one of them and they have not been supplied (e.g. because a KMS service does not have encrypt permissions) that func returns an error that is defined by the client.

---

Another alternative would be to combine the two by setting default encrypter/decrypter in the constructor that return an error and adding `Decryptor` and `Encryptor` funcs to override these with implementations that can successfully return.

Reply:
If you're going to go with functions that set the decryptor etc., consider using functional options ( http://dave.cheney.net/2014/10/17/functional-options-for-friendly-apis ), e.g. something like this:

```go
client := secrethub.NewClient(authMethod, secrethub.WithDecryptor(key), secrethub.WithEncryptor(key))

// package secrethub
type Option func(c *client) *client

func WithDecryptor(d Decryptor) Option {
   return func(c *client) *client {
      c.decryptor = d
      return c
   }
}
```

The code is quick, dirty and probably somewhat incorrect but you get the idea.

Reply:
What we should keep in mind that in practice a user of the client does not want to be concerned with what is used as and "encryptor" and what as a "decryptor". In practice the user provides one "method", e.g. "aws with KMS and STS" or "a credential file", and wants everything to just work. I think _we_ should figure out what consequences this has for the encryptor and decryptor, not the user creating the Client.

Reply:
Good point.

How about supplying an interface to the client that has methods to retrieve the encryptor, decryptor and authenticator? We could return such interfaces from the constructors, e.g. `NewCredential`, `NewAWSCredential` etc.

These can in turn return "fake" encryptors, that always return an error when using them.

For the user, usage can become something like this:

```
credential, err := secrethub.NewCredential("<your credential>", "<passphrase>")
client := secrethub.NewClient(credential, nil)
```
and when using AWS KMS:
```
credential, err := secrethub.NewAWSKMSCredential()
client := secrethub.NewClient(credential, nil)
```

Reply:
That sounds nice :+1:

Reply:
@SimonBarendse , what about this point?

Reply:
We leave this for now and pick it up later.

Reply:
Great insight @jpcoenen, it's really important to keep it simple for the user. Ideally, creating a `Client` should be a oneliner, including credential stuff. The more advanced use cases should be possible, but oneliners should be the default.